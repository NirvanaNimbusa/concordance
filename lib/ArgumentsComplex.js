'use strict'

const constants = require('./constants')
const Complex = require('./Complex')

class ArgumentsComplex extends Complex {
  constructor (tag, ctor, unwrapped, pointer, describeItem, describeProperty, instance) {
    super(tag, ctor, unwrapped, pointer, describeItem, describeProperty, instance)

    // Treat as an array, to allow comparisons with arrays
    this.instanceIsArray = true
    this.iteratorInstance = null
  }

  compare (expected) {
    const result = super.compare(expected)

    // Calling code may decide that Arguments can be compared to Arrays
    if (result === constants.UNEQUAL && expected.isComplex === true && expected.tag === 'Array') {
      return constants.AMBIGUOUS
    }

    return result
  }
}
Object.defineProperty(ArgumentsComplex.prototype, 'isArgumentsComplex', { value: true })
module.exports = ArgumentsComplex
