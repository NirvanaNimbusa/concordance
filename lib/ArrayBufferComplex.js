'use strict'

const Complex = require('./Complex')

const DEFAULT = {
  length: 0,
  next () { return null }
}

class ArrayBufferComplex extends Complex {
  createPropertyRecursor () {
    const recursor = super.createPropertyRecursor() || DEFAULT
    const length = recursor.length + 1

    let done = false
    const next = () => {
      if (done) return null

      const property = recursor.next()
      if (property) return property

      done = true

      return this.describeProperty('byteLength', this.instance.byteLength)
    }

    return { length, next }
  }

  createIterableRecursor () {
    const array = new Uint8Array(this.instance)
    const length = array.length

    let index = 0
    let iterator
    let done = false
    const next = () => {
      if (done) return null

      if (!iterator) {
        iterator = array[Symbol.iterator]()
      }

      const current = iterator.next()
      if (current.done) {
        done = true
      }

      const item = current.value
      if (done && item === undefined) return null

      return this.describeItem(index++, item)
    }

    return { length, next }
  }
}
module.exports = ArrayBufferComplex
