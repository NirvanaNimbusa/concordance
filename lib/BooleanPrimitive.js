'use strict'

const constants = require('./constants')

const DEEP_EQUAL = constants.DEEP_EQUAL
const UNEQUAL = constants.UNEQUAL

class BooleanPrimitive {
  constructor (value) {
    this.value = value
  }

  compare (expected) {
    return expected.isBooleanPrimitive === true && this.value === expected.value
      ? DEEP_EQUAL
      : UNEQUAL
  }
}
Object.defineProperty(BooleanPrimitive.prototype, 'isPrimitive', { value: true })
Object.defineProperty(BooleanPrimitive.prototype, 'isBooleanPrimitive', { value: true })
module.exports = BooleanPrimitive
