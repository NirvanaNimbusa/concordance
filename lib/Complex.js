'use strict'

const constants = require('./constants')
const getObjectKeys = require('./getObjectKeys')
const hasLength = require('./hasLength')

const DEEP_EQUAL = constants.DEEP_EQUAL
const SHALLOW_EQUAL = constants.SHALLOW_EQUAL
const UNEQUAL = constants.UNEQUAL

class IterableStats {
  constructor (recursor) {
    this.size = recursor.size
  }

  compare (expected) {
    return expected.isIterableStats === true && this.size === expected.size
      ? DEEP_EQUAL
      : UNEQUAL
  }
}
Object.defineProperty(IterableStats.prototype, 'isIterableStats', { value: true })

class ListStats {
  constructor (recursor) {
    this.size = recursor.size
  }

  compare (expected) {
    return expected.isListStats === true && this.size === expected.size
      ? DEEP_EQUAL
      : UNEQUAL
  }
}
Object.defineProperty(ListStats.prototype, 'isListStats', { value: true })

class PropertyStats {
  constructor (recursor) {
    this.size = recursor.size
  }

  compare (expected) {
    return expected.isPropertyStats === true && this.size === expected.size
      ? DEEP_EQUAL
      : UNEQUAL
  }
}
Object.defineProperty(PropertyStats.prototype, 'isPropertyStats', { value: true })

const DEFAULT_RECURSOR = {
  size: 0,
  next () { return null }
}

class Complex {
  constructor (
    tag, ctor, unwrapped, pointer, isArray, isIterable, isList,
    describeItem, describeMapEntry, describeProperty,
    instance
  ) {
    this.tag = tag
    this.ctor = ctor
    this.unwrapped = unwrapped
    this.pointer = pointer
    this.isArray = isArray
    this.isIterable = isIterable
    this.isList = isList

    this.describeItem = describeItem
    this.describeMapEntry = describeMapEntry
    this.describeProperty = describeProperty

    this.instance = instance
  }

  static fromValue (
    instance,
    tag, ctor, unwrapped, pointer,
    describeItem, describeMapEntry, describeProperty
  ) {
    const isArray = tag === 'Array'
    const isIterable = instance[Symbol.iterator] !== undefined
    const isList = isArray || hasLength(instance)

    return new this(
      tag, ctor, unwrapped, pointer, isArray, isIterable, isList,
      describeItem, describeMapEntry, describeProperty,
      instance
    )
  }

  createPropertyRecursor () {
    const objectKeys = getObjectKeys(this.instance, this.isList ? this.instance.length : 0)
    const size = objectKeys.size
    if (size === 0) return DEFAULT_RECURSOR

    let index = 0
    const next = () => {
      if (index === size) return null

      const key = objectKeys.keys[index++]
      return this.describeProperty(key, this.instance[key])
    }

    return { size, next }
  }

  createListRecursor () {
    if (!this.isList) return DEFAULT_RECURSOR

    const size = this.instance.length

    let index = 0
    const next = () => {
      if (index === size) return null

      const current = index
      index++
      return this.describeItem(current, this.instance[current])
    }

    return { size, next }
  }

  createIterableRecursor () {
    if (this.isArray || !this.isIterable) return DEFAULT_RECURSOR

    const iterator = this.instance[Symbol.iterator]()
    let first = iterator.next()

    let done = false
    let size = -1
    if (first.done) {
      if (first.value === undefined) {
        size = 0
        done = true
      } else {
        size = 1
      }
    }

    let index = 0
    const next = () => {
      if (done) return null

      while (!done) {
        const current = first || iterator.next()
        if (current === first) {
          first = null
        }
        if (current.done) {
          done = true
        }

        const item = current.value
        if (done && item === undefined) return null

        if (this.isList && this.instance[index] === item) {
          index++
        } else {
          return this.describeItem(index++, item)
        }
      }
    }

    return { size, next }
  }

  createRecursor () {
    let recursedProperty = false
    let recursedList = false
    let recursedIterable = false

    let recursor = null
    return () => {
      if (recursor !== null) {
        const retval = recursor.next()
        if (retval) {
          return retval
        } else {
          recursor = null
        }
      }

      // Prioritize recursing lists
      if (!recursedList && (this.isList || recursedProperty)) {
        recursor = this.createListRecursor()
        recursedList = true
        return new ListStats(recursor)
      }

      if (!recursedProperty) {
        recursor = this.createPropertyRecursor()
        recursedProperty = true
        return new PropertyStats(recursor)
      }

      if (!recursedIterable) {
        recursor = this.createIterableRecursor()
        recursedIterable = true
        return new IterableStats(recursor)
      }

      return null
    }
  }

  compare (expected) {
    if (expected.isComplex !== true) return UNEQUAL
    if (this.instance && expected.instance && this.instance === expected.instance) return DEEP_EQUAL
    if (this.tag !== expected.tag || this.ctor !== expected.ctor) return UNEQUAL
    if (this.unwrapped && expected.unwrapped && this.unwrapped.compare(expected.unwrapped) !== DEEP_EQUAL) return UNEQUAL
    return SHALLOW_EQUAL
  }
}
Object.defineProperty(Complex.prototype, 'isComplex', { value: true })
module.exports = Complex
