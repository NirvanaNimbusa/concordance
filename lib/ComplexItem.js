'use strict'

const constants = require('./constants')

const SHALLOW_EQUAL = constants.SHALLOW_EQUAL
const UNEQUAL = constants.UNEQUAL

class ComplexItem {
  constructor (index, value) {
    this.index = index
    this.value = value
  }

  createRecursor () {
    let done = false
    return () => {
      if (done) return null
      done = true
      return this.value
    }
  }

  compare (expected) {
    if (expected.isComplexItem !== true || this.index !== expected.index) return UNEQUAL

    const result = this.value.compare(expected.value)
    if (result !== UNEQUAL) return result

    if (this.value.isPointer === true) {
      return expected.value.isPointer === true
        ? UNEQUAL
        : SHALLOW_EQUAL
    }

    return expected.value.isPointer === true
      ? SHALLOW_EQUAL
      : UNEQUAL
  }

  format () {
    let value
    return {
      buffer (formatted) {
        value = formatted
      },
      finalize () {
        return value + ','
      }
    }
  }
}
Object.defineProperty(ComplexItem.prototype, 'isComplexItem', { value: true })
module.exports = ComplexItem
