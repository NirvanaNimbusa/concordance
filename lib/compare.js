'use strict'

const constants = require('./constants')
const describe = require('./describe')
const ArgumentsComplex = require('./ArgumentsComplex')
const Complex = require('./Complex')
const Pointer = require('./Pointer')

function compare (actual, expected) {
  if (actual === expected) return true

  const lhsLookup = new Map()
  const rhsLookup = new Map()
  const queue = [[describe(actual), describe(expected)]]
  while (queue.length > 0) {
    const pair = queue.shift()
    let lhs = pair[0]
    let rhs = pair[1]

    if (lhs instanceof Complex) {
      lhsLookup.set(lhs.pointer, lhs)
    }
    if (rhs instanceof Complex) {
      rhsLookup.set(rhs.pointer, rhs)
    }

    if (lhs instanceof Pointer) {
      if (rhs instanceof Pointer && lhs.compare(rhs) === constants.DEEP_EQUAL) continue

      lhs = lhsLookup.get(lhs.pointer)
    }
    if (rhs instanceof Pointer) {
      rhs = rhsLookup.get(rhs.pointer)
    }

    const result = lhs.compare(rhs)
    if (result === constants.UNEQUAL) return false
    if (result === constants.DEEP_EQUAL) continue

    const compareNested = result === constants.SHALLOW_EQUAL ||
      result === constants.AMBIGUOUS && lhs instanceof ArgumentsComplex
    if (!compareNested) return false

    const comparators = rhs[Symbol.iterator]()
    for (const comparable of lhs) {
      const step = comparators.next()
      if (step.done) return false

      queue.push([comparable, step.value])
    }

    if (!comparators.next().done) return false
  }

  return true
}
module.exports = compare
