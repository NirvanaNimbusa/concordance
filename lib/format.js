'use strict'

const describe = require('./describe')
const Circular = require('./Circular')

const alwaysFormat = () => true

function formatDescriptor (subject) {
  if (subject.isPrimitive === true) return subject.format('', '', '')

  const circular = new Circular()
  const lookup = new Map()

  let level
  let indent
  let innerIndent
  const setLevel = newLevel => {
    level = newLevel
    indent = '  '.repeat(level)
    innerIndent = indent + '  '
  }
  setLevel(0)

  const stack = []
  let topIndex = -1

  let retval
  while (retval === undefined) {
    if (subject.isComplex === true) {
      lookup.set(subject.pointer, subject)
    }

    if (subject.isPointer === true) {
      subject = lookup.get(subject.pointer)
    }

    let formatted
    if (circular.has(subject)) {
      formatted = '[Circular]'
    } else if (subject.format) {
      formatted = subject.format('', indent, innerIndent)
    }

    if (typeof formatted === 'string') {
      if (topIndex === -1) {
        retval = formatted
      } else {
        stack[topIndex].formatter.buffer('', formatted, subject)
      }
    } else {
      const record = {
        formatter: formatted,
        origin: subject,
        recursor: subject.createRecursor
          ? subject.createRecursor()
          : null,
        restoreLevel: level,
        shouldFormat: formatted.shouldFormat || alwaysFormat
      }

      if (record.formatter.nestInner) setLevel(level + 1)
      circular.add(subject)
      stack.push(record)
      topIndex++
    }

    while (retval === undefined && topIndex >= 0) {
      do {
        subject = stack[topIndex].recursor()
      } while (subject && !stack[topIndex].shouldFormat(subject))

      if (subject) {
        break
      }

      const record = stack.pop()
      topIndex--
      setLevel(record.restoreLevel)
      circular.delete(record.origin)

      const formattedRecord = record.formatter.finalize()
      if (formattedRecord !== null) {
        if (topIndex === -1) {
          retval = formattedRecord
        } else {
          stack[topIndex].formatter.buffer('', formattedRecord, record.origin)
        }
      } else {
        retval = null
      }
    }
  }

  return retval
}
exports.formatDescriptor = formatDescriptor

function format (value) {
  return formatDescriptor(describe(value))
}
exports.format = format
