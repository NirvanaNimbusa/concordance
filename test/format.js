import test from 'ava'

import {format} from '../lib/format'
import getStringTag from '../lib/getStringTag'

{
  const formatsPrimitive = (t, value) => t.snapshot(format(value))
  formatsPrimitive.title = (_, value) => {
    const str = String(value)
      .replace(/\r/g, '\\r')
      .replace(/\n/g, '\\n')
    return `formats primitive: ${str}`
  }
  for (const value of [
    null,
    undefined,
    false,
    true,
    '',
    'foo',
    '\\ -- \' -- "',
    'foo\nbar\\baz\'"',
    'qux\r\nquux',
    'qux\rquux',
    42,
    -42,
    0,
    -0,
    +0,
    Infinity,
    -Infinity,
    NaN,
    Symbol(), // eslint-disable-line symbol-description
    Symbol('foo'),
    Symbol.for('bar'),
    Symbol.iterator
  ]) {
    test(formatsPrimitive, value)
  }
}

test('formats a simple object', t => {
  const obj = { foo: 'bar', baz: 'qux' }
  const actual = format(obj)
  t.snapshot(actual)
})

test('formats a simple, nested object', t => {
  const obj = { foo: { baz: 'qux' } }
  const actual = format(obj)
  t.snapshot(actual)
})

test('formats multiline strings inside an object', t => {
  const actual = format({ 'foo\nbar': 'baz\nqux' })
  t.snapshot(actual)
})

{
  const formatsWrappedPrimitive = (t, value) => t.snapshot(format(Object(value)))
  formatsWrappedPrimitive.title = (_, value) => `formats wrapped primitive: ${String(value)}`
  for (const value of [
    false,
    true,
    42,
    -42,
    0,
    -0,
    +0,
    Infinity,
    -Infinity,
    NaN
  ]) {
    test(formatsWrappedPrimitive, value)
  }
}

test('formats wrapped strings as a list', t => {
  t.snapshot(format(Object('foo')))
})

test('formats a simple array', t => {
  const arr = ['foo', 'bar']
  const actual = format(arr)
  t.snapshot(actual)
})

test('formats a simple, nested array', t => {
  const arr = [['foo', 'bar']]
  const actual = format(arr)
  t.snapshot(actual)
})

test('formats an array with additional properties', t => {
  const arr = ['foo', 'bar']
  arr.baz = 'qux'
  const actual = format(arr)
  t.snapshot(actual)
})

test('formats a multiline string inside an array', t => {
  const actual = format(['bar\nbaz'])
  t.snapshot(actual)
})

test('formats maps', t => {
  const map = new Map()
  map.set('foo', 'bar')
  map.set({ baz: 'qux' }, 'quux')
  map.set('corge', { grault: 'garply' })
  const actual = format(map)
  t.snapshot(actual)
})

test('formats multiline strings inside a map', t => {
  const actual = format(new Map([['foo\nbar', 'baz\nqux']]))
  t.snapshot(actual)
})

test('formats maps with additional properties', t => {
  const map = new Map()
  map.set('foo', 'bar')
  map.baz = 'qux'
  const actual = format(map)
  t.snapshot(actual)
})

test('formats sets', t => {
  const set = new Set()
  set.add('foo')
  set.add({ bar: 'baz' })
  const actual = format(set)
  t.snapshot(actual)
})

test('formats a multiline string inside sets', t => {
  const actual = format(new Set(['bar\nbaz']))
  t.snapshot(actual)
})

test('formats sets with additional properties', t => {
  const set = new Set()
  set.add('foo')
  set.bar = 'baz'
  const actual = format(set)
  t.snapshot(actual)
})

test('formats funky objects that are lists and have an iterator', t => {
  const funky = {
    0: 'first',
    1: 'second',
    foo: 'bar'
  }
  Object.defineProperty(funky, 'length', { value: 2 })
  Object.defineProperty(funky, Symbol.iterator, { * value () { yield 'baz' } })

  const actual = format(funky)
  t.snapshot(actual)
})

test('formats regular expressions', t => {
  const actual = format(/foo/gim)
  t.snapshot(actual)
})

test('formats regular expressions with additional properties', t => {
  const actual = format(Object.assign(/foo/gim, { bar: 'baz' }))
  t.snapshot(actual)
})

test('formats anonymous functions', t => {
  const actual = format(() => {})
  t.snapshot(actual)
})

test('formats named functions', t => {
  const actual = format(function foo () {})
  t.snapshot(actual)
})

test('formats functions with additional properties', t => {
  const actual = format(Object.assign(function foo () {}, { bar: 'baz' }))
  t.snapshot(actual)
})

{
  // Node.js 4 provides Function, more recent versions use GeneratorFunction
  const tag = getStringTag(function * () {})

  test('formats anonymous generator functions', t => {
    const actual = format(function * () {})
    t.is(actual, `${tag} {}`)
  })

  test('formats named generator functions', t => {
    const actual = format(function * foo () {})
    t.is(actual, `${tag} foo {}`)
  })

  test('formats generator functions with additional properties', t => {
    const actual = format(Object.assign(function * foo () {}, { bar: 'baz' }))
    t.is(actual, `${tag} foo {
  bar: 'baz',
}`)
  })
}

test('formats arguments', t => {
  (function (a, b, c) {
    const actual = format(arguments)
    t.snapshot(actual)
  })('foo', 'bar', 'baz')
})

test('formats simple errors', t => {
  const actual = format(new TypeError('Test message'))
  t.snapshot(actual)
})

test('formats simple errors with a modified name', t => {
  const err = new TypeError('Test message')
  err.name = 'FooError'
  const actual = format(err)
  t.snapshot(actual)
})

test('formats errors with a name that does not include Error and does not match the constructor', t => {
  class Foo extends Error {
    constructor (message) {
      super(message)
      this.name = 'Bar'
    }
  }
  const actual = format(new Foo('Test message'))
  t.snapshot(actual)
})

test('formats errors with additional properties', t => {
  const actual = format(Object.assign(new TypeError('Test message'), { foo: 'bar' }))
  t.snapshot(actual)
})

test('formats promises', t => {
  const actual = format(Promise.resolve())
  t.snapshot(actual)
})

test('formats promises with additional properties', t => {
  const actual = format(Object.assign(Promise.resolve(), { foo: 'bar' }))
  t.snapshot(actual)
})

test('formats pointers', t => {
  const obj = { foo: 'bar' }
  const actual = format({ baz: obj, qux: { quux: obj } })
  t.snapshot(actual)
})

test('formats circular references', t => {
  const obj = {}
  obj.circular = obj
  const actual = format(obj)
  t.snapshot(actual)
})

{
  const plain = (t, value, tag) => {
    const actual = format(value)
    t.snapshot(actual)
  }
  plain.title = (_, __, tag) => `formats ${tag}`

  const withProperties = (t, value, tag) => {
    const actual = format(Object.assign(value, { foo: 'bar' }))
    t.snapshot(actual)
  }
  withProperties.title = (_, __, tag) => `formats ${tag} with additional properties`

  const buffer = Buffer.from('decafbad'.repeat(12), 'hex')
  const arrayBuffer = buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength)

  for (const [tag, value, valueForProps] of [
    ['ArrayBuffer', arrayBuffer],
    ['Buffer @Uint8Array', buffer],
    ['DataView', new DataView(arrayBuffer), new DataView(arrayBuffer)],
    ['Float32Array', new Float32Array(arrayBuffer)],
    ['Float64Array', new Float64Array(arrayBuffer)],
    ['Int16Array', new Int16Array(arrayBuffer)],
    ['Int32Array', new Int32Array(arrayBuffer)],
    ['Int8Array', new Int8Array(arrayBuffer)],
    ['Uint16Array', new Uint16Array(arrayBuffer)],
    ['Uint32Array', new Uint32Array(arrayBuffer)],
    ['Uint8Array', new Uint8Array(arrayBuffer)],
    ['Uint8ClampedArray', new Uint8ClampedArray(arrayBuffer)]
  ]) {
    test(plain, value, tag)
    test(withProperties, valueForProps || value.slice(), tag)
  }
}

test('formats dates', t => {
  const actual = format(new Date('1969-07-20T20:17:40Z'))
  t.snapshot(actual)
})

test('formats dates with additional properties', t => {
  const actual = format(Object.assign(new Date('1969-07-20T20:17:40Z'), { foo: 'bar' }))
  t.snapshot(actual)
})

test('shows non-Object tag if constructor name is different', t => {
  class Foo {}
  const actual1 = format(new Foo())
  t.snapshot(actual1)

  class Bar extends Array {}
  const actual2 = format(new Bar())
  t.snapshot(actual2)

  class Baz extends Date {}
  const actual3 = format(new Baz('1969-07-20T20:17:40Z'))
  t.snapshot(actual3)

  class Qux extends RegExp {}
  const actual4 = format(new Qux('foo'))
  t.snapshot(actual4)

  class Quux extends Int16Array {}
  const actual5 = format(new Quux())
  t.snapshot(actual5)
})

test('formats global', t => {
  t.snapshot(format(global))
})
