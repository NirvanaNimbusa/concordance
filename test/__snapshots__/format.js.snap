// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formats ArrayBuffer 1`] = `
"ArrayBuffer [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats ArrayBuffer with additional properties 1`] = `
"ArrayBuffer [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Buffer @Uint8Array 1`] = `
"Buffer @Uint8Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Buffer @Uint8Array with additional properties 1`] = `
"Buffer @Uint8Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats DataView 1`] = `
"DataView [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats DataView with additional properties 1`] = `
"DataView [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Float32Array 1`] = `
"Float32Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Float32Array with additional properties 1`] = `
"Float32Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Float64Array 1`] = `
"Float64Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Float64Array with additional properties 1`] = `
"Float64Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Int8Array 1`] = `
"Int8Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Int8Array with additional properties 1`] = `
"Int8Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Int16Array 1`] = `
"Int16Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Int16Array with additional properties 1`] = `
"Int16Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Int32Array 1`] = `
"Int32Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Int32Array with additional properties 1`] = `
"Int32Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Uint8Array 1`] = `
"Uint8Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Uint8Array with additional properties 1`] = `
"Uint8Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Uint8ClampedArray 1`] = `
"Uint8ClampedArray [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Uint8ClampedArray with additional properties 1`] = `
"Uint8ClampedArray [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Uint16Array 1`] = `
"Uint16Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Uint16Array with additional properties 1`] = `
"Uint16Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats Uint32Array 1`] = `
"Uint32Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
]"
`;

exports[`formats Uint32Array with additional properties 1`] = `
"Uint32Array [
  decafbad decafbad decafbad decafbad decafbad decafbad decafbad decafbad
  decafbad decafbad decafbad decafbad
  ---
  foo: 'bar',
]"
`;

exports[`formats a multiline string inside an array 1`] = `
"Array [
  \`bar␊
  baz\`,
]"
`;

exports[`formats a multiline string inside sets 1`] = `
"Set {
  \`bar␊
  baz\`,
}"
`;

exports[`formats a simple array 1`] = `
"Array [
  'foo',
  'bar',
]"
`;

exports[`formats a simple object 1`] = `
"Object {
  baz: 'qux',
  foo: 'bar',
}"
`;

exports[`formats a simple, nested array 1`] = `
"Array [
  Array [
    'foo',
    'bar',
  ],
]"
`;

exports[`formats a simple, nested object 1`] = `
"Object {
  foo: Object {
    baz: 'qux',
  },
}"
`;

exports[`formats an array with additional properties 1`] = `
"Array [
  'foo',
  'bar',
  ---
  baz: 'qux',
]"
`;

exports[`formats anonymous functions 1`] = `"Function {}"`;

exports[`formats arguments 1`] = `
"Arguments [
  'foo',
  'bar',
  'baz',
]"
`;

exports[`formats circular references 1`] = `
"Object {
  circular: [Circular],
}"
`;

exports[`formats dates 1`] = `"Date 1969-07-20T20:17:40.000Z {}"`;

exports[`formats dates with additional properties 1`] = `
"Date 1969-07-20T20:17:40.000Z {
  foo: 'bar',
}"
`;

exports[`formats errors with a name that does not include Error and does not match the constructor 1`] = `
"Bar (Foo) @Error {
  message: 'Test message',
}"
`;

exports[`formats errors with additional properties 1`] = `
"TypeError {
  foo: 'bar',
  message: 'Test message',
}"
`;

exports[`formats functions with additional properties 1`] = `
"Function foo {
  bar: 'baz',
}"
`;

exports[`formats funky objects that are lists and have an iterator 1`] = `
"Object [
  'first',
  'second',
  ---
  foo: 'bar',
  ---
  'baz',
]"
`;

exports[`formats global 1`] = `"Global {}"`;

exports[`formats maps 1`] = `
"Map {
  'foo' => 'bar',
  Object {
    baz: 'qux',
  } => 'quux',
  'corge' => Object {
    grault: 'garply',
  },
}"
`;

exports[`formats maps with additional properties 1`] = `
"Map {
  baz: 'qux',
  ---
  'foo' => 'bar',
}"
`;

exports[`formats multiline strings inside a map 1`] = `
"Map {
  \`foo␊
  bar\` => \`baz␊
  qux\`,
}"
`;

exports[`formats multiline strings inside an object 1`] = `
"Object {
  'foo\\\\nbar': \`baz␊
  qux\`,
}"
`;

exports[`formats named functions 1`] = `"Function foo {}"`;

exports[`formats pointers 1`] = `
"Object {
  baz: Object {
    foo: 'bar',
  },
  qux: Object {
    quux: Object {
      foo: 'bar',
    },
  },
}"
`;

exports[`formats primitive:  1`] = `"''"`;

exports[`formats primitive: \\ -- ' -- " 1`] = `"'\\\\\\\\ -- \\\\' -- \\"'"`;

exports[`formats primitive: -42 1`] = `"-42"`;

exports[`formats primitive: -Infinity 1`] = `"-Infinity"`;

exports[`formats primitive: 0 1`] = `"0"`;

exports[`formats primitive: 0 2`] = `"-0"`;

exports[`formats primitive: 0 3`] = `"0"`;

exports[`formats primitive: 42 1`] = `"42"`;

exports[`formats primitive: Infinity 1`] = `"Infinity"`;

exports[`formats primitive: NaN 1`] = `"NaN"`;

exports[`formats primitive: Symbol() 1`] = `"Symbol()"`;

exports[`formats primitive: Symbol(Symbol.iterator) 1`] = `"Symbol.iterator"`;

exports[`formats primitive: Symbol(bar) 1`] = `"Symbol(bar)"`;

exports[`formats primitive: Symbol(foo) 1`] = `"Symbol(foo)"`;

exports[`formats primitive: false 1`] = `"false"`;

exports[`formats primitive: foo 1`] = `"'foo'"`;

exports[`formats primitive: foo\\nbar\\baz'" 1`] = `
"\`foo␊
bar\\\\\\\\baz'\\"\`"
`;

exports[`formats primitive: null 1`] = `"null"`;

exports[`formats primitive: qux\\r\\nquux 1`] = `
"\`qux␍␊
quux\`"
`;

exports[`formats primitive: qux\\rquux 1`] = `
"\`qux␍
quux\`"
`;

exports[`formats primitive: true 1`] = `"true"`;

exports[`formats primitive: undefined 1`] = `"undefined"`;

exports[`formats promises 1`] = `"Promise {}"`;

exports[`formats promises with additional properties 1`] = `
"Promise {
  foo: 'bar',
}"
`;

exports[`formats regular expressions 1`] = `"/foo/gim"`;

exports[`formats regular expressions with additional properties 1`] = `
"RegExp {
  /foo/gim
  ---
  bar: 'baz',
}"
`;

exports[`formats sets 1`] = `
"Set {
  'foo',
  Object {
    bar: 'baz',
  },
}"
`;

exports[`formats sets with additional properties 1`] = `
"Set {
  bar: 'baz',
  ---
  'foo',
}"
`;

exports[`formats simple errors 1`] = `
"TypeError {
  message: 'Test message',
}"
`;

exports[`formats simple errors with a modified name 1`] = `
"FooError (TypeError) {
  message: 'Test message',
}"
`;

exports[`formats wrapped primitive: -42 1`] = `
"Number {
  -42
}"
`;

exports[`formats wrapped primitive: -Infinity 1`] = `
"Number {
  -Infinity
}"
`;

exports[`formats wrapped primitive: 0 1`] = `
"Number {
  0
}"
`;

exports[`formats wrapped primitive: 0 2`] = `
"Number {
  -0
}"
`;

exports[`formats wrapped primitive: 0 3`] = `
"Number {
  0
}"
`;

exports[`formats wrapped primitive: 42 1`] = `
"Number {
  42
}"
`;

exports[`formats wrapped primitive: Infinity 1`] = `
"Number {
  Infinity
}"
`;

exports[`formats wrapped primitive: NaN 1`] = `
"Number {
  NaN
}"
`;

exports[`formats wrapped primitive: false 1`] = `
"Boolean {
  false
}"
`;

exports[`formats wrapped primitive: true 1`] = `
"Boolean {
  true
}"
`;

exports[`formats wrapped strings as a list 1`] = `
"String [
  'f',
  'o',
  'o',
]"
`;

exports[`shows non-Object tag if constructor name is different 1`] = `"Foo {}"`;

exports[`shows non-Object tag if constructor name is different 2`] = `"Bar @Array []"`;

exports[`shows non-Object tag if constructor name is different 3`] = `"Baz @Date 1969-07-20T20:17:40.000Z {}"`;

exports[`shows non-Object tag if constructor name is different 4`] = `
"Qux @RegExp {
  /foo/
}"
`;

exports[`shows non-Object tag if constructor name is different 5`] = `"Quux @Int16Array []"`;
